### PR TITLE
Fix column types when default column is being patched for missing data features

### PR DESCRIPTION
**Pull request by rakeshkashyap123**
Fix column types when default column is being patched for missing data features in both anchored and SWA features.

### CODE REVIEW

File: feathr-impl/src/main/scala/com/linkedin/feathr/offline/swa/SlidingWindowAggregationJoiner.scala [hunk 1]
```diff
@@ -281,14 +281,14 @@ private[offline] class SlidingWindowAggregationJoiner(
...
-        val userSpecifiedTypesConfig = windowAggAnchorDFThisStage.flatMap(_._1.featureAnchor.featureTypeConfigs)
+        val userSpecifiedTypesConfig = windowAggAnchorDFMap.flatMap(_._1.featureAnchor.featureTypeConfigs)
 
         // Create a map from the feature name to the column format, ie - RAW or FDS_TENSOR
         val featureNameToColumnFormat = allWindowAggFeatures.map (nameToFeatureAnchor => nameToFeatureAnchor._1 ->
           nameToFeatureAnchor._2.featureAnchor.extractor
           .asInstanceOf[TimeWindowConfigurableAnchorExtractor].features(nameToFeatureAnchor._1).columnFormat)
 
-        contextDF = emptyFeatures.foldLeft(contextDF) { (contextDF, featureName) =>
+        contextDF = emptyFeatures.intersect(joinedFeatures.toSet).foldLeft(contextDF) { (contextDF, featureName) =>
```
Comment (anirudhagar13) [inline feathr-impl/src/main/scala/com/linkedin/feathr/offline/swa/SlidingWindowAggregationJoiner.scala:291]: If I understand correctly, this solves the problem of not adding empty features that are going to be skipped right?

File: feathr-impl/src/main/scala/com/linkedin/feathr/offline/swa/SlidingWindowAggregationJoiner.scala [hunk 2]
```diff
@@ -281,14 +281,14 @@ private[offline] class SlidingWindowAggregationJoiner(
           val features = x._1.selectedFeatures
           emptyFeatures.contains(features.head)
         }).flatMap(s => s._1.featureAnchor.defaults) // populate the defaults for features whose data was missing
-        val userSpecifiedTypesConfig = windowAggAnchorDFThisStage.flatMap(_._1.featureAnchor.featureTypeConfigs)
+        val userSpecifiedTypesConfig = windowAggAnchorDFMap.flatMap(_._1.featureAnchor.featureTypeConfigs)
```
Comment (anirudhagar13) [inline feathr-impl/src/main/scala/com/linkedin/feathr/offline/swa/SlidingWindowAggregationJoiner.scala:284]: Does this solve the problem of getting proper featureType configs from the original dataframe, which has inferred or specific type for empty features, contrary to not getting them and having to use default string -> float?

File: feathr-impl/src/main/scala/com/linkedin/feathr/offline/join/workflow/AnchoredFeatureJoinStep.scala [hunk 3]
```diff
@@ -108,20 +98,21 @@ private[offline] class AnchoredFeatureJoinStep(
...
-      substituteDefaultsForDataMissingFeatures(ctx.sparkSession, observationDF, ctx.logicalPlan,
-        missingAnchoredFeatures)
-    } else observationDF
+      (substituteDefaultsForDataMissingFeatures(ctx.sparkSession, observationDF, ctx.logicalPlan,
+        missingAnchoredFeatures), missingFeatures)
+    } else (observationDF, Seq())
 
     val allAnchoredFeatures: Map[String, FeatureAnchorWithSource] = ctx.featureGroups.allAnchoredFeatures
-    val joinStages = ctx.logicalPlan.joinStages
+    val joinStages = ctx.logicalPlan.joinStages.map(keyValue => (keyValue._1, keyValue._2.filter(featuresAtThisStage =>
```
Comment (anirudhagar13) [inline feathr-impl/src/main/scala/com/linkedin/feathr/offline/join/workflow/AnchoredFeatureJoinStep.scala:114]: Unsure what is happening here, might need to connect offline, but is the idea here to skip joinStages for those features that are missing?

Does it solve a new problem for us? Or shouldn't this have caused an issue before?

**Review by anirudhagar13 — COMMENTED** [review]
Some nit comments / request for additional unit tests.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 4]
```diff
@@ -137,10 +137,11 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
       |  multiply_a_b: "toNumeric(aa) * toNumeric(bb)"
       |
       |  categorical_b: {
-      |    key: [foo]
-      |    inputs: { foo_b: { key: foo, feature: bb } }
-      |    definition: "toCategorical(foo_b)"
-      |  }
+      |      key: [foo]
+      |      inputs: { foo_b: { key: foo, feature: bb } }
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:141]: There are changes in this file.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 5]
```diff
@@ -547,10 +547,10 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
     assertEquals(featureList(0).getAs[Row]("aEmbedding"),
       Row(mutable.WrappedArray.make(Array("")), mutable.WrappedArray.make(Array(2.0f))))
     assertEquals(featureList(0).getAs[Row]("featureWithNull3"), "null")
-    assertEquals(featureList(0).getAs[Row]("featureWithNull5"), mutable.Map("" -> 1.0f))
+    assertEquals(featureList(0).getAs[Row]("featureWithNull5"), null)
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:550]: My bad, the @Test annotation was removed from this test, so the test was not being run. I have fixed the values of the columns according to the default type which is generated.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 6]
```diff
@@ -547,10 +547,10 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
...
     assertEquals(featureList(0).getAs[Row]("featureWithNull7"), null)
     assertEquals(featureList(0).getAs[Row]("featureWithNull"),-1.0f)
-    assertEquals(featureList(0).getAs[Row]("featureWithNull4"),Map())
+    assertEquals(featureList(0).getAs[Row]("featureWithNull4"), null)
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:553]: Fixed.
The format of the data actually enforces the type. So, I dont think we need to check the type of column again.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 7]
```diff
@@ -547,10 +547,10 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
...
     assertEquals(featureList(0).getAs[Row]("featureWithNull2"),1.0f)
     assertEquals(featureList(0).getAs[Row]("derived_featureWithNull"),
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:555]: My bad, the @test annotation was removed from this test, so the test was not being run.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala [hunk 8]
```diff
@@ -1055,7 +1055,7 @@ class SlidingWindowAggIntegTest extends FeathrIntegTest {
     res.show()
     val df = res.collect()(0)
     assertEquals(df.getAs[Float]("simplePageViewCount"), 10f)
-    assertEquals(df.getAs[Float]("simpleFeature"), Row(mutable.WrappedArray.make(Array("")), mutable.WrappedArray.make(Array(20.0f))))
+    assertEquals(df.getAs[Float]("simpleFeature"), 20.0f)
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala:1058]: "featureWithNull5" is an example.

**Review by anirudhagar13 — APPROVED** [review]
LGTM!

**Pull request merged**